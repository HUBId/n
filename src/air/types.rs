//! Core type definitions for the AIR layer.
//!
//! The final API surface is still under construction; concrete data
//! structures will be introduced as the proof system stabilises.

use core::fmt;

use crate::field::prime_field::CanonicalSerialize;
use crate::field::FieldElement as Felt;
use crate::utils::serialization::FieldElementBytes;

/// Column identifier newtype used when reporting errors.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct ColIx(usize);

impl ColIx {
    /// Creates a new column index wrapper.
    pub const fn new(ix: usize) -> Self {
        Self(ix)
    }

    /// Returns the underlying index as `usize`.
    pub const fn as_usize(self) -> usize {
        self.0
    }
}

impl fmt::Display for ColIx {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "#{}", self.0)
    }
}

/// Describes the role a trace column plays in the execution trace.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TraceRole {
    /// Main trace columns feed directly into constraint evaluation.
    Main,
    /// Auxiliary columns are derived from witness relations.
    Auxiliary,
}

impl fmt::Display for TraceRole {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TraceRole::Main => f.write_str("main"),
            TraceRole::Auxiliary => f.write_str("auxiliary"),
        }
    }
}

/// Little-endian domain extension order.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LdeOrder {
    blowup: usize,
}

impl LdeOrder {
    /// Creates a new LDE order descriptor.
    pub fn new(blowup: usize) -> Result<Self, AirError> {
        if blowup == 0 {
            return Err(AirError::LayoutViolation(
                "lde blowup factor must be strictly positive",
            ));
        }
        if !blowup.is_power_of_two() {
            return Err(AirError::LayoutViolation(
                "lde blowup factor must be a power of two",
            ));
        }
        Ok(Self { blowup })
    }

    /// Returns the configured blowup factor.
    pub const fn blowup(&self) -> usize {
        self.blowup
    }

    fn validate_rows(&self, rows: usize) -> Result<(), AirError> {
        if rows == 0 {
            return Err(AirError::LayoutViolation(
                "execution trace must contain at least one row",
            ));
        }
        if rows % self.blowup != 0 {
            return Err(AirError::LayoutViolation(
                "lde blowup factor must divide the execution trace length",
            ));
        }
        Ok(())
    }
}

/// Degree bounds for trace and auxiliary polynomials.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DegreeBounds {
    main: usize,
    auxiliary: usize,
}

impl DegreeBounds {
    /// Constructs new degree bounds, enforcing consistency between roles.
    pub fn new(main: usize, auxiliary: usize) -> Result<Self, AirError> {
        if main == 0 {
            return Err(AirError::LayoutViolation(
                "main trace degree bound must be positive",
            ));
        }
        if auxiliary < main {
            return Err(AirError::LayoutViolation(
                "auxiliary degree bound must be at least the main bound",
            ));
        }
        Ok(Self { main, auxiliary })
    }

    fn check_role(&self, role: TraceRole, observed_degree: usize) -> Result<(), AirError> {
        let bound = match role {
            TraceRole::Main => self.main,
            TraceRole::Auxiliary => self.auxiliary,
        };
        if observed_degree > bound {
            Err(AirError::DegreeOverflow {
                role,
                observed: observed_degree,
                allowed: bound,
            })
        } else {
            Ok(())
        }
    }

    /// Ensures the observed degree stays within the configured bounds for `role`.
    pub fn ensure(&self, role: TraceRole, observed_degree: usize) -> Result<(), AirError> {
        self.check_role(role, observed_degree)
    }
}

/// Boundary position within the execution trace.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BoundaryAt {
    /// First row of the trace.
    First,
    /// Last row of the trace.
    Last,
    /// Arbitrary row index.
    Row(usize),
}

impl fmt::Display for BoundaryAt {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BoundaryAt::First => f.write_str("first"),
            BoundaryAt::Last => f.write_str("last"),
            BoundaryAt::Row(ix) => write!(f, "row {ix}"),
        }
    }
}

/// Metadata describing a single trace column.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceColMeta {
    /// Human readable column label.
    pub label: String,
    /// The role played by this column.
    pub role: TraceRole,
    /// Boundary values that are exposed publicly or enforced statically.
    pub boundaries: Vec<BoundaryAt>,
}

impl TraceColMeta {
    /// Creates a new trace column metadata descriptor.
    pub fn new(label: impl Into<String>, role: TraceRole) -> Self {
        Self {
            label: label.into(),
            role,
            boundaries: Vec::new(),
        }
    }

    /// Registers an additional boundary location.
    pub fn with_boundary(mut self, boundary: BoundaryAt) -> Self {
        if !self.boundaries.contains(&boundary) {
            self.boundaries.push(boundary);
        }
        self
    }

    fn allows(&self, boundary: &BoundaryAt) -> bool {
        self.boundaries.contains(boundary)
    }
}

/// Public input specification derived from the trace schema.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublicSpec {
    /// Fields exposed to the verifier.
    pub fields: Vec<PublicFieldMeta>,
}

impl PublicSpec {
    /// Validates the specification against a trace schema.
    pub fn validate(&self, schema: &TraceSchema) -> Result<(), AirError> {
        for field in &self.fields {
            if let PublicFieldType::TraceValue { column, boundary } = &field.field_type {
                let meta = schema
                    .column_meta(*column)
                    .ok_or_else(|| AirError::SchemaMismatch {
                        what: "trace column count",
                        expected: schema.columns.len(),
                        actual: column.as_usize() + 1,
                    })?;
                if !meta.allows(boundary) {
                    return Err(AirError::BoundaryViolation {
                        column: *column,
                        boundary: *boundary,
                        detail: "boundary not declared in schema",
                    });
                }
            }
        }
        Ok(())
    }
}

/// Metadata for a single public field.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublicFieldMeta {
    /// Identifier used in documentation and transcripts.
    pub label: String,
    /// Typing information for the exposed field.
    pub field_type: PublicFieldType,
}

/// Public field typing.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PublicFieldType {
    /// Constant value published independently of the trace.
    Constant(Felt),
    /// Value sourced from the execution trace.
    TraceValue { column: ColIx, boundary: BoundaryAt },
}

/// AIR level execution trace schema.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceSchema {
    /// Metadata for each execution trace column.
    pub columns: Vec<TraceColMeta>,
    /// Domain extension order.
    pub lde_order: LdeOrder,
    /// Degree bounds for trace polynomials.
    pub degree_bounds: DegreeBounds,
}

impl TraceSchema {
    /// Creates a new trace schema descriptor.
    pub fn new(
        columns: Vec<TraceColMeta>,
        lde_order: LdeOrder,
        degree_bounds: DegreeBounds,
    ) -> Result<Self, AirError> {
        if columns.is_empty() {
            return Err(AirError::SchemaMismatch {
                what: "trace column count",
                expected: 1,
                actual: 0,
            });
        }
        Ok(Self {
            columns,
            lde_order,
            degree_bounds,
        })
    }

    /// Returns metadata for the requested column index.
    pub fn column_meta(&self, ix: ColIx) -> Option<&TraceColMeta> {
        self.columns.get(ix.as_usize())
    }

    /// Validates an execution trace against the schema.
    pub fn validate_trace(&self, trace: &TraceData) -> Result<(), AirError> {
        if self.columns.len() != trace.num_columns() {
            return Err(AirError::SchemaMismatch {
                what: "trace column count",
                expected: self.columns.len(),
                actual: trace.num_columns(),
            });
        }
        self.lde_order.validate_rows(trace.num_rows())?;
        let observed_degree = trace.num_rows().saturating_sub(1);
        for (ix, meta) in self.columns.iter().enumerate() {
            self.degree_bounds.check_role(meta.role, observed_degree)?;
            for boundary in &meta.boundaries {
                trace.boundary_value(ColIx::new(ix), *boundary)?;
            }
        }
        Ok(())
    }
}

/// Execution trace container used for schema validation and serialization.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceData {
    columns: Vec<Vec<Felt>>,
    rows: usize,
}

impl TraceData {
    /// Creates a new trace data container.
    pub fn new(columns: Vec<Vec<Felt>>) -> Result<Self, AirError> {
        if columns.is_empty() {
            return Err(AirError::SchemaMismatch {
                what: "trace column count",
                expected: 1,
                actual: 0,
            });
        }
        let rows = columns[0].len();
        if rows == 0 {
            return Err(AirError::LayoutViolation(
                "execution trace must contain at least one row",
            ));
        }
        for column in &columns {
            if column.len() != rows {
                return Err(AirError::SchemaMismatch {
                    what: "trace row count",
                    expected: rows,
                    actual: column.len(),
                });
            }
        }
        Ok(Self { columns, rows })
    }

    /// Returns the number of columns.
    pub fn num_columns(&self) -> usize {
        self.columns.len()
    }

    /// Returns the number of rows.
    pub fn num_rows(&self) -> usize {
        self.rows
    }

    /// Returns a reference to a trace column.
    pub fn column(&self, ix: ColIx) -> Result<&[Felt], AirError> {
        self.columns
            .get(ix.as_usize())
            .map(|col| col.as_slice())
            .ok_or_else(|| AirError::SchemaMismatch {
                what: "trace column count",
                expected: self.columns.len(),
                actual: ix.as_usize() + 1,
            })
    }

    /// Returns the trace value at the specified boundary.
    pub fn boundary_value(&self, ix: ColIx, boundary: BoundaryAt) -> Result<Felt, AirError> {
        let column = self.column(ix)?;
        let row = match boundary {
            BoundaryAt::First => 0,
            BoundaryAt::Last => column.len().saturating_sub(1),
            BoundaryAt::Row(r) => r,
        };
        column.get(row).copied().ok_or(AirError::BoundaryViolation {
            column: ix,
            boundary,
            detail: "boundary outside trace length",
        })
    }

    /// Serialises a column into canonical little-endian field element bytes.
    pub fn column_bytes(&self, ix: ColIx) -> Result<Vec<FieldElementBytes>, AirError> {
        let column = self.column(ix)?;
        let mut bytes = Vec::with_capacity(column.len());
        for value in column {
            let mut buf = [0u8; 32];
            let le = value.to_bytes().map_err(|_| AirError::Serialization {
                kind: SerKind::Trace,
                detail: "non-canonical trace value",
            })?;
            buf[..le.len()].copy_from_slice(&le);
            bytes.push(FieldElementBytes { bytes: buf });
        }
        Ok(bytes)
    }
}

/// Serialization context kinds surfaced in [`AirError::Serialization`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SerKind {
    /// Execution trace serialisation.
    Trace,
    /// Public input serialisation.
    PublicInput,
    /// Transcript serialisation.
    Transcript,
}

impl fmt::Display for SerKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SerKind::Trace => f.write_str("trace data"),
            SerKind::PublicInput => f.write_str("public input"),
            SerKind::Transcript => f.write_str("transcript"),
        }
    }
}

/// Error enumeration covering deterministic AIR failures.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AirError {
    /// Schema layout mismatch (column count, row count, etc.).
    SchemaMismatch {
        /// Description of the component that failed validation.
        what: &'static str,
        /// Expected cardinality.
        expected: usize,
        /// Actual cardinality encountered.
        actual: usize,
    },
    /// Schema declared a boundary outside the trace layout.
    InvalidBoundary {
        /// Column where the invalid boundary was declared.
        column: ColIx,
        /// Boundary location that failed validation.
        boundary: BoundaryAt,
        /// Additional diagnostic context.
        detail: &'static str,
    },
    /// Boundary constraint violation.
    BoundaryViolation {
        /// Column where the violation occurred.
        column: ColIx,
        /// Boundary location.
        boundary: BoundaryAt,
        /// Additional diagnostic context.
        detail: &'static str,
    },
    /// Transition relation failed for the provided witness values.
    InvalidTransition {
        /// Column identifier.
        column: ColIx,
        /// Step index.
        step: usize,
        /// Value at `step`.
        current: Felt,
        /// Value at `step + 1`.
        next: Felt,
    },
    /// Degree bound exceeded for the specified role.
    DegreeOverflow {
        /// Role for which the overflow was detected.
        role: TraceRole,
        /// Observed degree.
        observed: usize,
        /// Allowed maximum degree.
        allowed: usize,
    },
    /// Serialization failure for one of the helper contexts.
    Serialization {
        /// Serialization context.
        kind: SerKind,
        /// Description of the failure.
        detail: &'static str,
    },
    /// Witness column admitted non-determinism.
    NonDeterministicWitness {
        /// Column identifier.
        column: ColIx,
        /// Step index where mismatch was detected.
        step: usize,
    },
    /// General layout violation (e.g. invalid parameters).
    LayoutViolation(&'static str),
}

impl fmt::Display for AirError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AirError::SchemaMismatch {
                what,
                expected,
                actual,
            } => {
                write!(
                    f,
                    "schema mismatch for {what}: expected {expected}, got {actual}"
                )
            }
            AirError::InvalidBoundary {
                column,
                boundary,
                detail,
            } => write!(
                f,
                "invalid boundary at column {column} ({boundary}): {detail}"
            ),
            AirError::BoundaryViolation {
                column,
                boundary,
                detail,
            } => write!(
                f,
                "boundary violation at column {column} ({boundary}): {detail}"
            ),
            AirError::InvalidTransition {
                column,
                step,
                current,
                next,
            } => write!(
                f,
                "invalid transition at column {column} step {step}: {} -> {}",
                current.0, next.0
            ),
            AirError::DegreeOverflow {
                role,
                observed,
                allowed,
            } => write!(
                f,
                "degree overflow for {role} columns: observed {observed}, allowed {allowed}"
            ),
            AirError::Serialization { kind, detail } => {
                write!(f, "serialization error for {kind}: {detail}")
            }
            AirError::NonDeterministicWitness { column, step } => write!(
                f,
                "non-deterministic witness detected at column {column} step {step}"
            ),
            AirError::LayoutViolation(detail) => write!(f, "layout violation: {detail}"),
        }
    }
}

impl std::error::Error for AirError {}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_schema() -> Result<TraceSchema, AirError> {
        let columns = vec![
            TraceColMeta::new("main", TraceRole::Main)
                .with_boundary(BoundaryAt::First)
                .with_boundary(BoundaryAt::Last),
            TraceColMeta::new("aux", TraceRole::Auxiliary).with_boundary(BoundaryAt::Row(1)),
        ];
        let lde = LdeOrder::new(4)?;
        let degree = DegreeBounds::new(3, 4)?;
        TraceSchema::new(columns, lde, degree)
    }

    fn sample_trace() -> Result<TraceData, AirError> {
        TraceData::new(vec![
            vec![
                Felt::from(1u64),
                Felt::from(2u64),
                Felt::from(3u64),
                Felt::from(4u64),
            ],
            vec![
                Felt::from(5u64),
                Felt::from(6u64),
                Felt::from(7u64),
                Felt::from(8u64),
            ],
        ])
    }

    #[test]
    fn schema_validation_accepts_matching_trace() -> Result<(), AirError> {
        let schema = sample_schema()?;
        let trace = sample_trace()?;
        schema.validate_trace(&trace)?;
        Ok(())
    }

    #[test]
    fn schema_validation_detects_mismatched_columns() -> Result<(), AirError> {
        let schema = sample_schema()?;
        let trace = TraceData::new(vec![vec![
            Felt::from(1u64),
            Felt::from(2u64),
            Felt::from(3u64),
            Felt::from(4u64),
        ]])?;
        let result = schema.validate_trace(&trace);
        match result {
            Err(AirError::SchemaMismatch {
                what,
                expected,
                actual,
            }) => {
                if what != "trace column count" {
                    return Err(AirError::LayoutViolation("unexpected mismatch kind"));
                }
                if expected != 2 || actual != 1 {
                    return Err(AirError::LayoutViolation("unexpected mismatch values"));
                }
                Ok(())
            }
            Err(err) => Err(err),
            Ok(_) => Err(AirError::LayoutViolation("schema should reject trace")),
        }
    }

    #[test]
    fn trace_serialization_uses_little_endian_helpers() -> Result<(), AirError> {
        let trace = TraceData::new(vec![
            vec![Felt::from(1u64), Felt::from(0x0102_0304_0506_0708u64)],
            vec![Felt::from(9u64), Felt::from(10u64)],
        ])?;
        let column_bytes = trace.column_bytes(ColIx::new(0))?;
        if column_bytes.len() != 2 {
            return Err(AirError::LayoutViolation(
                "unexpected column serialization length",
            ));
        }
        if column_bytes[0].bytes[..8] != 1u64.to_le_bytes() {
            return Err(AirError::LayoutViolation("first element not little-endian"));
        }
        if column_bytes[0].bytes[8..].iter().any(|b| *b != 0) {
            return Err(AirError::LayoutViolation("first element padding not zero"));
        }
        if column_bytes[1].bytes[..8] != 0x0102_0304_0506_0708u64.to_le_bytes() {
            return Err(AirError::LayoutViolation(
                "second element not little-endian",
            ));
        }
        Ok(())
    }

    #[test]
    fn air_error_formatting_is_human_readable() -> Result<(), AirError> {
        let transition = AirError::InvalidTransition {
            column: ColIx::new(1),
            step: 3,
            current: Felt::from(5u64),
            next: Felt::from(7u64),
        };
        if format!("{transition}") != "invalid transition at column #1 step 3: 5 -> 7" {
            return Err(AirError::LayoutViolation(
                "unexpected transition formatting",
            ));
        }

        let serialization = AirError::Serialization {
            kind: SerKind::Trace,
            detail: "io failure",
        };
        if format!("{serialization}") != "serialization error for trace data: io failure" {
            return Err(AirError::LayoutViolation(
                "unexpected serialization formatting",
            ));
        }
        Ok(())
    }
}
