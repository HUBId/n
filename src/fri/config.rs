//! Canonical configuration descriptions for the FRI protocol.
//! Provides named profiles that are shared between prover and verifier.

/// Canonical digest representation for a FRI parameter set.
///
/// The digest is computed over the canonical serialization of the
/// folding factor, folding mode, query schedule, and target depth
/// using BLAKE3.
/// The hexadecimal strings exposed below allow integrators to pin
/// protocol parameters without embedding executable verification logic.
pub type ParameterDigest = &'static str;

/// Describes a complete FRI profile that can be referenced by the
/// STARK configuration and transcript builders.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FriProfile {
    /// Human readable profile identifier.
    pub name: &'static str,
    /// Folding factor applied to each layer; we expose the canonical binary
    /// folding schedule in this module.
    pub folding_factor: usize,
    /// Number of sampling queries performed by the verifier.
    pub query_count: usize,
    /// Target depth (logarithmic size) of the final layer after folding.
    pub target_depth: usize,
    /// Digest binding the profile to a concrete transcript layout.
    pub parameter_digest: ParameterDigest,
}

/// Standard security profile targeting typical rollup deployments.
pub const STANDARD_FRI_PROFILE: FriProfile = FriProfile {
    name: "standard",
    folding_factor: 2,
    query_count: 64,
    target_depth: 6,
    parameter_digest: "5570706d57436b4fa198d0e8c8a6d3fa01d8c348c602248b066ee8f3fe0eb566",
};

/// High security profile with an increased query budget.
pub const HISEC_FRI_PROFILE: FriProfile = FriProfile {
    name: "hisec",
    folding_factor: 2,
    query_count: 96,
    target_depth: 8,
    parameter_digest: "40b923845a41a78da7bdef951f541ae1b6b30794b0244a54b19fc5801d686b06",
};
