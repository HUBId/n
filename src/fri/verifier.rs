use super::folding::{coset_shift_schedule, next_domain_size, parent_index};
use super::layer::verify_query_opening;
use super::proof::{derive_query_positions, hash_final_layer, FriProof};
use super::types::{FriError, FriSecurityLevel};
use super::{fe_mul, fe_sub, hash_leaf, BINARY_FOLD_ARITY};
use crate::field::prime_field::FieldElementOps;
use crate::field::FieldElement;
use crate::hash::merkle::EMPTY_DIGEST;
use crate::params::StarkParams;
use crate::transcript::{Transcript, TranscriptError, TranscriptLabel};
use crate::utils::serialization::DigestBytes;

fn map_transcript_error(err: TranscriptError) -> FriError {
    match err {
        TranscriptError::InvalidLabel => FriError::InvalidStructure("transcript-label"),
        TranscriptError::RangeZero => FriError::InvalidStructure("transcript-range"),
        TranscriptError::Overflow => FriError::InvalidStructure("transcript-overflow"),
        TranscriptError::Serialization(_) => FriError::InvalidStructure("transcript-ser"),
        TranscriptError::BoundsViolation => FriError::InvalidStructure("transcript-bounds"),
        TranscriptError::Unsupported => FriError::InvalidStructure("transcript-unsupported"),
    }
}

fn security_level_from_query_count(count: usize) -> Result<FriSecurityLevel, FriError> {
    if count == FriSecurityLevel::Standard.query_budget() {
        Ok(FriSecurityLevel::Standard)
    } else if count == FriSecurityLevel::HiSec.query_budget() {
        Ok(FriSecurityLevel::HiSec)
    } else if count == FriSecurityLevel::Throughput.query_budget() {
        Ok(FriSecurityLevel::Throughput)
    } else {
        Err(FriError::InvalidStructure("unsupported-query-count"))
    }
}

pub fn fri_verify(
    proof: &FriProof,
    params: &StarkParams,
    transcript: &mut Transcript,
) -> Result<(), FriError> {
    if proof.initial_domain_size == 0 {
        return Err(FriError::EmptyCodeword);
    }

    let expected_domain_size = 1usize << (params.fri().domain_log2 as usize);
    if proof.initial_domain_size != expected_domain_size {
        return Err(FriError::InvalidStructure("initial-domain-size"));
    }

    let query_count = params.fri().queries as usize;
    let security_level = security_level_from_query_count(query_count)?;
    if proof.security_level != security_level {
        return Err(FriError::SecurityLevelMismatch);
    }

    if proof.queries.len() != query_count {
        return Err(FriError::QueryBudgetMismatch {
            expected: query_count,
            actual: proof.queries.len(),
        });
    }

    let num_layers = params.fri().num_layers as usize;
    if proof.layer_roots.len() != num_layers {
        return Err(FriError::InvalidStructure("layer-count"));
    }
    if proof.fold_challenges.len() != num_layers {
        return Err(FriError::InvalidStructure("fold challenge length"));
    }

    let residual_bound = params.fri().r as usize;
    if proof.final_polynomial.len() > residual_bound {
        return Err(FriError::InvalidStructure("final-polynomial-length"));
    }

    let recomputed_digest = hash_final_layer(&proof.final_polynomial);
    if recomputed_digest != proof.final_polynomial_digest {
        return Err(FriError::LayerRootMismatch { layer: num_layers });
    }

    let coset_shifts = coset_shift_schedule(params, num_layers);

    for (layer_index, root) in proof.layer_roots.iter().enumerate() {
        if layer_index > u8::MAX as usize {
            return Err(FriError::InvalidStructure("layer-index-overflow"));
        }
        transcript
            .absorb_digest(
                TranscriptLabel::FriRoot(layer_index as u8),
                &DigestBytes { bytes: *root },
            )
            .map_err(map_transcript_error)?;
        let eta = transcript
            .challenge_field(TranscriptLabel::FriFoldChallenge(layer_index as u8))
            .map_err(map_transcript_error)?;
        let expected = proof
            .fold_challenges
            .get(layer_index)
            .copied()
            .ok_or(FriError::InvalidStructure("missing-fold-challenge"))?;
        if eta != expected {
            return Err(FriError::InvalidStructure("fold challenge mismatch"));
        }
    }

    let count_bytes = (query_count as u32).to_le_bytes();
    transcript
        .absorb_bytes(TranscriptLabel::QueryCount, &count_bytes)
        .map_err(map_transcript_error)?;
    let query_seed_bytes = transcript
        .challenge_bytes(TranscriptLabel::QueryIndexStream, 32)
        .map_err(map_transcript_error)?;
    let mut query_seed = [0u8; 32];
    query_seed.copy_from_slice(&query_seed_bytes);

    let positions = derive_query_positions(query_seed, query_count, proof.initial_domain_size);
    if positions.len() != proof.queries.len() {
        return Err(FriError::InvalidStructure("query count mismatch"));
    }

    for (expected_position, query) in positions.iter().zip(proof.queries.iter()) {
        if query.position != *expected_position {
            return Err(FriError::InvalidStructure("query position mismatch"));
        }
        if query.layers.len() != num_layers {
            return Err(FriError::InvalidStructure("layer count mismatch"));
        }

        let mut index = *expected_position;
        let mut domain_size = proof.initial_domain_size;
        for (layer_index, coset_shift) in coset_shifts.iter().take(num_layers).enumerate() {
            let opening = &query.layers[layer_index];
            let root = &proof.layer_roots[layer_index];
            verify_query_opening(layer_index, opening, root, index, domain_size)?;

            let first = opening
                .path
                .first()
                .ok_or(FriError::InvalidStructure("merkle-path-empty"))?;
            let child_position = first.index.0 as usize;
            if child_position >= BINARY_FOLD_ARITY {
                return Err(FriError::InvalidStructure("merkle-index"));
            }
            if index % BINARY_FOLD_ARITY != child_position {
                return Err(FriError::InvalidStructure("pair-index-mismatch"));
            }

            let beta = proof.fold_challenges[layer_index];
            let parent_value = if layer_index + 1 < num_layers {
                query.layers[layer_index + 1].value
            } else {
                query.final_value
            };

            let sibling_digest = first.siblings[0];
            match child_position {
                0 => {
                    let diff = fe_sub(parent_value, opening.value);
                    let beta_shift = fe_mul(beta, *coset_shift);
                    if beta_shift == FieldElement::ZERO {
                        if diff != FieldElement::ZERO {
                            return Err(FriError::FoldingConstraintViolated { layer: layer_index });
                        }
                    } else {
                        let inv = beta_shift
                            .inv()
                            .ok_or(FriError::FoldingConstraintViolated { layer: layer_index })?;
                        let sibling_value = fe_mul(diff, inv);
                        if sibling_digest == EMPTY_DIGEST {
                            if sibling_value != FieldElement::ZERO {
                                return Err(FriError::FoldingConstraintViolated {
                                    layer: layer_index,
                                });
                            }
                        } else if hash_leaf(&sibling_value) != sibling_digest {
                            return Err(FriError::FoldingConstraintViolated { layer: layer_index });
                        }
                    }
                }
                1 => {
                    let beta_shift = fe_mul(beta, *coset_shift);
                    let scaled = fe_mul(beta_shift, opening.value);
                    let left_value = fe_sub(parent_value, scaled);
                    if sibling_digest == EMPTY_DIGEST {
                        if left_value != FieldElement::ZERO {
                            return Err(FriError::FoldingConstraintViolated { layer: layer_index });
                        }
                    } else if hash_leaf(&left_value) != sibling_digest {
                        return Err(FriError::FoldingConstraintViolated { layer: layer_index });
                    }
                }
                _ => unreachable!(),
            }

            index = parent_index(index);
            domain_size = next_domain_size(domain_size);
        }

        if index >= proof.final_polynomial.len() {
            return Err(FriError::QueryOutOfRange {
                position: *expected_position,
            });
        }

        if proof.final_polynomial[index] != query.final_value {
            return Err(FriError::LayerRootMismatch { layer: num_layers });
        }
    }

    Ok(())
}
