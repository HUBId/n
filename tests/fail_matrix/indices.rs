use insta::assert_debug_snapshot;
use rpp_stark::config::ProofKind as ConfigProofKind;
use rpp_stark::proof::types::VerifyError;
use rpp_stark::proof::verifier::verify;

use super::{
    duplicate_composition_index, duplicate_trace_index, mismatch_composition_indices,
    mismatch_trace_indices, swap_composition_indices, swap_trace_indices, FailMatrixFixture,
};

#[test]
fn trace_rejects_unsorted_indices() {
    let fixture = FailMatrixFixture::new();
    let mutated = swap_trace_indices(&fixture.proof());

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    );

    let error = report.error.expect("expected verification failure");
    assert!(matches!(error, VerifyError::IndicesNotSorted));

    assert_debug_snapshot!(
        "trace_rejects_unsorted_indices",
        mutated.proof.openings().trace().indices()
    );
}

#[test]
fn trace_rejects_duplicate_indices() {
    let fixture = FailMatrixFixture::new();
    let mutated = duplicate_trace_index(&fixture.proof());

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    );

    let error = report.error.expect("expected verification failure");
    assert!(matches!(error, VerifyError::IndicesDuplicate { .. }));

    assert_debug_snapshot!(
        "trace_rejects_duplicate_indices",
        mutated.proof.openings().trace().indices()
    );
}

#[test]
fn trace_rejects_mismatched_indices() {
    let fixture = FailMatrixFixture::new();
    let mutated = mismatch_trace_indices(&fixture.proof());

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    );

    let error = report.error.expect("expected verification failure");
    assert!(matches!(error, VerifyError::IndicesMismatch));

    assert_debug_snapshot!(
        "trace_rejects_mismatched_indices",
        mutated.proof.openings().trace().indices()
    );
}

#[test]
fn composition_rejects_unsorted_indices() {
    let fixture = FailMatrixFixture::new();
    let Some(mutated) = swap_composition_indices(&fixture.proof()) else {
        eprintln!("fixture does not expose composition openings; skipping test");
        return;
    };

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    );

    let error = report.error.expect("expected verification failure");
    assert!(matches!(error, VerifyError::IndicesNotSorted));

    assert_debug_snapshot!(
        "composition_rejects_unsorted_indices",
        mutated
            .proof
            .openings()
            .composition()
            .expect("composition openings available")
            .indices()
    );
}

#[test]
fn composition_rejects_duplicate_indices() {
    let fixture = FailMatrixFixture::new();
    let Some(mutated) = duplicate_composition_index(&fixture.proof()) else {
        eprintln!("fixture does not expose composition openings; skipping test");
        return;
    };

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    );

    let error = report.error.expect("expected verification failure");
    assert!(matches!(error, VerifyError::IndicesDuplicate { .. }));

    assert_debug_snapshot!(
        "composition_rejects_duplicate_indices",
        mutated
            .proof
            .openings()
            .composition()
            .expect("composition openings available")
            .indices()
    );
}

#[test]
fn composition_rejects_mismatched_indices() {
    let fixture = FailMatrixFixture::new();
    let Some(mutated) = mismatch_composition_indices(&fixture.proof()) else {
        eprintln!("fixture does not expose composition openings; skipping test");
        return;
    };

    let public_inputs = fixture.public_inputs();
    let config = fixture.config();
    let context = fixture.verifier_context();

    let report = verify(
        ConfigProofKind::Tx,
        &public_inputs,
        &mutated.bytes,
        &config,
        &context,
    );

    let error = report.error.expect("expected verification failure");
    assert!(matches!(error, VerifyError::IndicesMismatch));

    assert_debug_snapshot!(
        "composition_rejects_mismatched_indices",
        mutated
            .proof
            .openings()
            .composition()
            .expect("composition openings available")
            .indices()
    );
}
