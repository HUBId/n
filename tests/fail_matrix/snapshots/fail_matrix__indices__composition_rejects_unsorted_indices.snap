---
source: tests/fail_matrix/indices.rs
assertion_line: 116
expression: "&mutated.proof.openings.composition.as_ref().expect(\"composition openings available\").indices"
---
[
    26,
    3,
    42,
    45,
    59,
    74,
    117,
    118,
    127,
    140,
    172,
    174,
    191,
    197,
    213,
    214,
    234,
    243,
    266,
    276,
    324,
    344,
    371,
    373,
    377,
    384,
    397,
    412,
    433,
    455,
    471,
    479,
    492,
    522,
    555,
    572,
    577,
    584,
    592,
    604,
    621,
    670,
    671,
    696,
    700,
    729,
    734,
    739,
    745,
    746,
    757,
    772,
    813,
    819,
    822,
    825,
    840,
    900,
    914,
    956,
    977,
    1015,
    1019,
    1021,
]
